### PR TITLE
Prefix scan, iterator changes, multi-get

### PR DESCRIPTION
**Pull request by edmundnoble**

### CODE REVIEW

File: src/Data/CAS/RocksDB.hs [hunk 1]
```diff
@@ -231,25 +247,27 @@ resetOpenRocksDb path = do
...
     return db
   where
-    opts = modernDefaultOptions { R.createIfMissing = True, R.errorIfExists = True }
+    opts = modernDefaultOptions { R.errorIfExists = True }
 
 -- | Close a 'RocksDb' instance.
 --
 closeRocksDb :: RocksDb -> IO ()
-closeRocksDb = R.close . _rocksDbHandle
+closeRocksDb (RocksDb (R.DB db_ptr) _) = C.c_rocksdb_close db_ptr
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:255]: What is the issue with the original code. The new version seems to access APIs that should be internal to the Haskell rocksdb package.

File: src/Data/CAS/RocksDB.hs [hunk 2]
```diff
@@ -191,24 +197,34 @@ instance NoThunks RocksDb where
...
-  where
-    opts = modernDefaultOptions { R.createIfMissing = True }
+openRocksDb :: FilePath -> C.OptionsPtr -> IO RocksDb
+openRocksDb path opts_ptr = do
+    GHC.setFileSystemEncoding GHC.utf8
+    createDirectoryIfMissing True path
+    db <- withFilePath path $ \path_ptr ->
+        liftM R.DB
+        $ R.throwIfErr "open"
+        $ C.c_rocksdb_open opts_ptr path_ptr
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:220]: Why can't we use the more high-level `R.open` call here as in the original version?

File: src/Data/CAS/RocksDB.hs [hunk 3]
```diff
@@ -191,24 +197,34 @@ instance NoThunks RocksDb where
...
 -- at the provided directory path, a new database is created.
 --
-openRocksDb :: FilePath -> IO RocksDb
-openRocksDb path = do
-    db <- RocksDb <$> R.open path opts <*> mempty
-    initializeRocksDb db
-    return db
-  where
-    opts = modernDefaultOptions { R.createIfMissing = True }
+openRocksDb :: FilePath -> C.OptionsPtr -> IO RocksDb
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:213]: `C.OptionsPtr` seems very low-level. This module exposes three different representations of options at the API level:

* `C.OptionsPtr`,
* `R.Options'`, and
* `R.Options`.

I think, the first two should be avoided on API level. The first one should probably be used only in the Haskell rocksdb package.

File: src/Data/CAS/RocksDB.hs [hunk 4]
```diff
@@ -231,25 +247,27 @@ resetOpenRocksDb path = do
     initializeRocksDb db
     return db
   where
-    opts = modernDefaultOptions { R.createIfMissing = True, R.errorIfExists = True }
+    opts = modernDefaultOptions { R.errorIfExists = True }
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:250]: Why is `resetOpenRocksDb` not exposing the options, while `openRocksDb` and `withRocksDb` do it?

File: src/Data/CAS/RocksDB.hs [hunk 5]
```diff
@@ -191,24 +197,34 @@ instance NoThunks RocksDb where
...
+        liftM R.DB
+        $ R.throwIfErr "open"
+        $ C.c_rocksdb_open opts_ptr path_ptr
+    let rdb = RocksDb db mempty
+    initializeRocksDb rdb
+    return rdb
+
+withOpts :: R.Options -> (R.Options' -> IO a) -> IO a
+withOpts opts =
+    bracket (R.mkOpts opts) R.freeOpts
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:227]: Is it possible to manage the lifetime of options through a foreign ptr that is retained by the RocksDb value on the Haskell Heap that uses the options?

File: src/Data/CAS/RocksDB.hs [hunk 6]
```diff
@@ -421,7 +452,7 @@ data RocksDbTableIter k v = RocksDbTableIter
     { _rocksDbTableIterValueCodec :: !(Codec v)
     , _rocksDbTableIterKeyCodec :: !(Codec k)
     , _rocksDbTableIterNamespace :: !B.ByteString
-    , _rocksDbTableIter :: !I.Iterator
+    , _rocksDbTableIter :: !C.IteratorPtr
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs]: Why do we have to go that low-level here and can't continue to use `I.Iterator`?

File: src/Data/CAS/RocksDB.hs [hunk 7]
```diff
@@ -435,125 +466,112 @@ instance NoThunks (RocksDbTableIter k v) where
     {-# INLINE wNoThunks #-}
     {-# INLINE showTypeOf #-}
 
--- | Creates a 'RocksDbTableIterator'. If the 'RocksDbTable' is not empty, the
--- iterator is pointing to the first key in the 'RocksDbTable' and is valid.
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:439]: This comment should be retained. Maybe by moving it to `withTableIter`. 

There are lots of semantic invariants attached to iterators that aren't reflected in the types. Thus these functions should be documented very carefully.

File: src/Data/CAS/RocksDB.hs [hunk 8]
```diff
@@ -435,125 +466,112 @@ instance NoThunks (RocksDbTableIter k v) where
...
 -- A valid iterator returns a value when 'tableIterEntry', 'tableIterValue', or
 -- 'tableIterKey' is called on it.
 --
-tableIterValid :: MonadIO m => RocksDbTableIter k v -> m Bool
-tableIterValid it = I.iterKey (_rocksDbTableIter it) >>= \case
-    Nothing -> return False
-    (Just !x) -> return $! checkIterKey it x
+tableIterValid :: RocksDbTableIter k v -> IO Bool
+tableIterValid it =
+    I.iterValid (_rocksDbTableIter it)
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:505]: Can you explain why this change maintains the original semantics of the function? It's not immediately obvious. I assume it has to do with the new `readOptions` that define the bounds for the iterator?

File: chainweb-storage.cabal [hunk 9]
```diff
@@ -47,13 +42,14 @@ library
...
         , directory >=1.3
+        , filepath
         , deepseq >=1.4
         , exceptions >=0.10
         , hashable >=1.2
         , lens >=4.16
         , memory >=0.14
         , nothunks >= 0.1.0.0
-        , rocksdb-haskell >=1.0.1
+        , rocksdb-haskell-kadena >= 1.1.0
```
Comment (larskuhtz) [inline chainweb-storage.cabal]:
```suggestion
        , rocksdb-haskell-kadena >=1.1.0
```

**Review by larskuhtz — COMMENTED** [review]
Some questions from a preliminary review. Mostly about API abstraction levels and semantics of iterators.

File: src/Data/CAS/RocksDB.hs (hunk 1 above)
Comment (edmundnoble) [reply src/Data/CAS/RocksDB.hs:255]: Double-frees the options, unfortunately. Will be dealt with in the follow-up, which will clean up the options ptr handling.

File: src/Data/CAS/RocksDB.hs (hunk 2 above)
Comment (edmundnoble) [reply src/Data/CAS/RocksDB.hs:220]: Will be dealt with with the options ptr cleanup, no way to use `R.open` without it allocating the options itself.

File: src/Data/CAS/RocksDB.hs (hunk 3 above)
Comment (edmundnoble) [reply src/Data/CAS/RocksDB.hs:213]: Ditto.

File: src/Data/CAS/RocksDB.hs (hunk 5 above)
Comment (edmundnoble) [reply src/Data/CAS/RocksDB.hs:227]: Yeah something like a foreign ptr would be fine, we'll see how it ends up.

File: src/Data/CAS/RocksDB.hs (hunk 8 above)
Comment (edmundnoble) [reply src/Data/CAS/RocksDB.hs:505]: Exactly. It just so happens that rocksdb's iterator bounds have the exact same semantics as our table bounds want: the iterator stops being valid if it seeks outside of the range (according to `I.iterValid`), `iterLast` and `iterFirst` respect the range, and a new iterator starts at `iterFirst`.
